examples/vm_scheduling/reinforcement_learning/common/components/state.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import numpy as np
+import pandas as pd
+
+
+PM_ATTRIBUTES = ["cpu_cores_capacity", "memory_capacity", \
+                 "cpu_cores_allocated", "memory_allocated"]
+
+VM_ATTRIBUTES = ["cpu_cores_requirement", "memory_requirement", "lifetime", "remain_time", "total_income"]
+
+
+class VMState(object):
+    def __init__(
+        self, pm_num, durations, vm_state_path, vm_window_size, pm_window_size
+    ):
+        self._pm_num = pm_num
+        self._durations = durations
+        self._vm_states = np.load(vm_state_path)
+        self._dim = (pm_num * 2) * pm_window_size + len(VM_ATTRIBUTES) * vm_window_size
+
+        self._history_pm_state = np.zeros((pm_window_size - 1, self._pm_num, 2))
+        self._st, self._vm_window_size, self._pm_window_size = 0, vm_window_size, pm_window_size
+
+    def __call__(self, env, decision_event):
+        # get state
+        pm_state = self._get_pm_state(env)
+        vm_state = self._get_vm_state(env, decision_event)
+        state = np.concatenate((pm_state.flatten(), vm_state.flatten()))
+        # get legal pm
+        legal_pm = self._get_legal_pm(decision_event, pm_state)
+
+        return state, legal_pm
+
+    def _get_pm_state(self, env):
+        total_pm_info = env.snapshot_list["pms"][
+            env.frame_index::PM_ATTRIBUTES
+        ].reshape(self._pm_num, len(PM_ATTRIBUTES))
+
+        # normalize the attributes of pms' cpu and memory
+        self._max_cpu_capacity = np.max(total_pm_info[:, 0])
+        self._max_memory_capacity = np.max(total_pm_info[:, 1])
+        total_pm_info[:, 2] /= self._max_cpu_capacity
+        total_pm_info[:, 3] /= self._max_memory_capacity
+
+        # get the remaining cpu and memory of the pms
+        remain_cpu = (1 - total_pm_info[:, 2]).reshape(1, self._pm_num, 1)
+        remain_memory = (1 - total_pm_info[:, 3]).reshape(1, self._pm_num, 1)
+
+        # get the pms' information
+        total_pm_info = np.concatenate((remain_cpu, remain_memory), axis=2)
+
+        # get the sequence pms' information
+        self._history_pm_state = np.concatenate((self._history_pm_state, total_pm_info), axis=0)
```

Comment:
keep all the history?

Reply:
The history_pm_state will store the total history information. Although some history information will never be used, it's difficult to remove the numpy data, so I choose store them all.